charms/k8s/src/charm.py

```diff
@@ -108,53 +130,25 @@ def _install_k8s_snap(self):
             channel = self.config["channel"]
             k8s_snap.ensure(SnapState.Latest, channel=channel)
 
-    def _on_install(self, event):
-        """Handle install event for the charm.
-
-        Args:
-            event: The event that triggered this handler.
-        """
-        # TODO: Implement clustering using leader units.
-        try:
-            self._install_k8s_snap()
-            self._apply_snap_requirements()
-            self._bootstrap_k8s_snap()
-            self._enable_components()
-            self.unit.status = ops.WaitingStatus("Waiting for K8s to be ready")
-        except InvalidResponseError as e:
-            logger.warning("Failed to query k8s snap. Reason: %s", e)
-            self.unit.status = ops.WaitingStatus("Waiting for K8sd API.")
-            event.defer()
-        except SnapError as e:
-            logger.error("Failed to install k8s snap. Reason: %s", e.message)
-            self.unit.status = ops.BlockedStatus("Failed to install k8s snap")
-            event.defer()
-        except subprocess.CalledProcessError as e:
-            logger.error("Failed to run subprocess: %s", e)
-            event.defer()
-        except K8sdConnectionError as e:
-            logger.warning("Unable to contact K8sd API: %s", e)
-            self.unit.status = ops.WaitingStatus("Waiting for K8sd API.")
-            event.defer()
-
-    def _on_update_status(self, event: ops.UpdateStatusEvent):
-        """Handle update-status event.
-
-        Args:
-            event: event triggering the handler.
-        """
+    def _update_status(self):
+        """Check k8s snap status."""
         try:
             if self.api_manager.is_cluster_ready():
                 if version := self._get_snap_version():
                     self.unit.set_workload_version(version)
-                self.unit.status = ops.ActiveStatus("Ready")
+            else:
+                status.add(ops.WaitingStatus("Waiting for k8s to be ready."))
         except InvalidResponseError:
-            self.unit.status = ops.WaitingStatus("Waiting for K8s to be ready")
-            event.defer()
+            log.info("Invalid Response from K8sd...")
+            status.add(ops.WaitingStatus("Waiting for K8s to be ready"))
```

Comment:
This will stay in the `WaitingStatus` forever if there is a bug in `k8sd`. Should we change into `ErrorStatus` after some retries?

Reply:
Correct. Personally, I believe the charm must be agnostic (at certain level) to the underlying machinery (i.e., k8s-snap). Our CI machinery only passes the build/deploy stage if the charm becomes active/idle. Handling this retry might involve implementing a busy-waiting mechanism or storing the state, which is not the best option in certain cases. Could we improve the snap side to expose a mechanism that flags when the snap is unable to come up?